File: algorithms/graphs/complete_graph_missing_edges.rs
// cargo-deps: itertools
extern crate itertools;
use itertools::Itertools;

fn missing_edges(n: i32, edges: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let all_edges = (0..n).combinations(2);
    all_edges.filter(|e| !edges.contains(e) && !edges.contains(&vec![e[1], e[0]]) )
             .collect()
}

fn main() {
    let edges = vec![vec![0,1], vec![1,2], vec![2,0]];
    let missing = missing_edges(4, edges);
    println!("{:?}", missing);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn trivial_graph_test() {
        let n = 1;
        let edges: Vec<Vec<i32>> = vec![];
        let actual = missing_edges(n, edges);
        let expected: Vec<Vec<i32>> = vec![];
        assert_eq!(actual, expected);
    }

    #[test]
    fn no_missing_edges_test() {
        let n = 3;
        let edges = vec![vec![0, 1], vec![1, 2], vec![0,2]];
        let actual = missing_edges(n, edges);
        let expected: Vec<Vec<i32>> = vec![];
        assert_eq!(actual, expected);
    }

    #[test]
    fn some_missing_edges_test() {
        let n = 4;
        let edges = vec![vec![0, 1], vec![1, 2], vec![2,0]];
        let actual = missing_edges(n, edges);
        let expected = vec![vec![0, 3], vec![1, 3], vec![2,3]];
        assert_eq!(actual, expected);
    }
}

// how to run
// $ cargo install cargo-script
// $ cargo script complete_graph_missing_edges.rs

// how to test:
// $ cargo script complete_graph_missing_edges.rs --test

File: algorithms/graphs/complete_graph_missing_edges.rb
# frozen_string_literal: true

require 'test/unit/assertions'
include Test::Unit::Assertions

# given `n` as the number of nodes in a network
# and `edges` as the list of edges connecting two nodes
# find the list of missing edges to be added to the graph
# to make it a complete undirected graph
# where every node is connected to all the others

def missing_edges(n, edges)
  all_edges = (0...n).to_a.combination(2).to_a
  all_edges - edges.map(&:sort)
end

# trivial graph
def trivial_graph_test
  n = 1
  edges = []
  actual = missing_edges(n, edges)
  expected = []
  assert_equal(actual, expected)
end

# no missing edges
def no_missing_edges_test
  n = 3
  edges = [[0, 1], [1, 2], [0, 2]]
  actual = missing_edges(n, edges)
  expected = []

  assert_equal actual, expected
end

# some missing edges
def some_missing_edges_test
  n = 4
  edges = [[0, 1], [1, 2], [2, 0]]
  actual = missing_edges(n, edges)
  expected = [[0, 3], [1, 3], [2, 3]]

  assert_equal(actual, expected)
end

# run test suite

trivial_graph_test
no_missing_edges_test
some_missing_edges_test
